src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,207 @@ MutableBigInteger sqrt() {
+     *
+     * @return the integer square root of {@code this} and the remainder if needed
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+            long s = ulongSqrt(x);
+
+            return new MutableBigInteger[] {
+                    new MutableBigInteger((int) s),
+                    needRemainder ? new MutableBigInteger(x - s * s) : null
+            };
+        }
+
+        long bitLength = this.bitLength();
+        if (bitLength != (int) bitLength)
+            throw new ArithmeticException("bitLength() integer overflow");
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.primitiveLeftShift(shift);
+        }
+
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder && !sqrtRem[1].isZero()) {
+                final int s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0) { // An optimization
+                    // s0 is at most 15 bit long
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul(s0 << 1, doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
+                }
+                sqrtRem[1].primitiveRightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
+
+    private static long ulongSqrt(long x) {
+        /* For every long value s in [0, 2^32) such that x == s * s,
+         * it is true that s - 1 <= (long) Math.sqrt(x >= 0 ? x : x + 0x1p64) <= s,
+         * and if x == 2^64 - 1, then (long) Math.sqrt(x >= 0 ? x : x + 0x1p64) == 2^32.
+         * This means that the value returned by Math.sqrt()
+         * for a long value in the range [0, 2^64) is either correct,
+         * or rounded up/down by one if the value is too high
+         * and too close to a perfect square.
+         */
+        long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
+        if (s > LONG_MASK // avoid overflow of s * s
+                || Long.compareUnsigned(x, s * s) < 0) {
+            s--;
+        } else if (s < LONG_MASK) { // avoid overflow of (s + 1) * (s + 1)
+            long s1 = s + 1;
+            if (Long.compareUnsigned(x, s1 * s1) >= 0)
+                s = s1;
+        }
+        return s;
```

Comment:
Experimentally, the following seems a bit faster.
In some cases, it avoids a full multiplication, some updates, and has one less test.
I hope it is correct as well ;-)

```suggestion
        long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
        long s2 = s * s;  // overflows iff s = 2^32
        if (s == 1L << 32
                || Long.compareUnsigned(x, s2) < 0) {
            return s - 1;
        }
        if (Long.compareUnsigned(x, s2 + 2 * s) <= 0) {  // x < (s + 1)^2
            return s;
        }
        return s + 1;
```

Reply:
> Experimentally, the following seems a bit faster. In some cases, it avoids a full multiplication, some updates, and has one less test. I hope it is correct as well ;-)

It's nice, but I'm afraid that if `s == LONG_MASK` and `Long.compareUnsigned(x, s * s) >= 0`, the overflow check is unavoidable...

Reply:
I guess you are concerned about an overflow in `s2 + 2 * s`?

If s = 2^32 - 1, then s2 = 2^64 - 2·2^32 + 1 and 2s = 2·2^32 - 2, without overflows.
Thus, s2 + 2s = 2^64 - 1, without overflows.

Reply:
@rgiulietti True, it's almost borderline code...